File: 129/sum-root-to-leaf-numbers.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int sumNumbers(TreeNode* root) {
        int sum = 0;
        return allChildrenValue(root, 0);
    }
    int allChildrenValue(TreeNode* now, int parentValue) {
        int nowValue = parentValue*10+now->val;
        if(now->left==nullptr && now->right==nullptr) return nowValue;
        if(now->left==nullptr) return allChildrenValue(now->right, nowValue);
        if(now->right==nullptr) return allChildrenValue(now->left, nowValue);
        return allChildrenValue(now->left, nowValue) + allChildrenValue(now->right, nowValue);
    }
};

File: 129/sum-root-to-leaf-numbers.py
# 把root走到leaf的數值，全部加起來，簡單「函式呼叫函式」即可！
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sumNumbers(self, root: Optional[TreeNode]) -> int:
        def pathSum(root, prev): # prev代表「之前累積的數」
            if root==None: return 0
            if root.left==None and root.right==None:
                return root.val + prev*10 # 遇到葉子，就可以回傳
            return pathSum(root.left, root.val+prev*10) + pathSum(root.right, root.val+prev*10)
        return pathSum(root, 0)
